web/src/settings_org.js

```diff
@@ -702,29 +706,191 @@ export function discard_property_element_changes(elem, for_realm_default_setting
         });
     }
 
-    const html_body = render_settings_deactivate_realm_modal();
+    const time_unit_choices = ["days", "weeks"];
+    const minimum_allowed_days = page_params.min_deactivated_realm_deletion_days;
+    const maximum_allowed_days = page_params.max_deactivated_realm_deletion_days;
+
+    const delete_in_options = {};
+
+    function get_allowed_duration_message() {
+        if (maximum_allowed_days === null) {
+            return $t(
+                {
+                    defaultMessage:
+                        "Enter a value greater than or equal to {minimum_allowed_days} days",
+                },
+                {minimum_allowed_days},
+            );
+        }
+        return $t(
+            {
+                defaultMessage:
+                    "Enter a value between {minimum_allowed_days} and {maximum_allowed_days} days",
+            },
+            {minimum_allowed_days, maximum_allowed_days},
+        );
+    }
+
+    function is_valid_time_period(time_period) {
+        if (time_period === "custom") {
+            return true;
+        }
+        if (time_period === "null") {
+            if (maximum_allowed_days === null) {
+                return true;
+            }
+            return false;
+        }
+        if (maximum_allowed_days === null && time_period >= minimum_allowed_days * 24 * 60) {
+            return true;
+        }
+        if (
+            time_period >= minimum_allowed_days * 24 * 60 &&
+            time_period <= maximum_allowed_days * 24 * 60
+        ) {
+            return true;
+        }
+        return false;
+    }
+
+    for (const key in settings_config.realm_deletion_in_values) {
+        if (is_valid_time_period(settings_config.realm_deletion_in_values[key].value)) {
+            delete_in_options[key] = settings_config.realm_deletion_in_values[key];
+        }
+    }
+
+    const html_body = render_settings_deactivate_realm_modal({
+        is_admin: page_params.is_admin,
+        is_owner: page_params.is_owner,
+        development_environment: page_params.development_environment,
+        delete_in_options,
+        allowed_duration_message: get_allowed_duration_message(),
+        time_choices: time_unit_choices,
+    });
+
+    function deactivate_realm_modal_post_render() {
+        settings_ui.set_custom_time_inputs_visibility(
+            $("#deletes_in"),
+            custom_deletion_time_unit,
+            custom_deletion_time_input,
+        );
+
+        let deletion_time_in_minutes;
+        if ($("#deletes_in").val() !== "custom") {
+            deletion_time_in_minutes = $("#deletes_in").val();
+        } else {
+            deletion_time_in_minutes = settings_ui.get_custom_input_time_in_minutes(
+                custom_deletion_time_unit,
+                custom_deletion_time_input,
+            );
+        }
+        const valid_to_text = valid_to(deletion_time_in_minutes);
+        settings_ui.set_valid_till_text($("#deletes_in"), valid_to_text);
```

Comment:
This code is duplicated and is used two times in invite code apart form being used twice here. Would have been good to have a common function doing this but `valid_to` is different for invite case, so I am not sure. We can probably pass `valid_to` as an argument to the common function. Let's wait for more feedback on this.

Reply:
Probably best to leave this as a follow-up; I agree it's messy but I'm not sure what the best fix is.